File: packages/pyre/flow/Product.py
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis
# orthologue
# (c) 1998-2019 all rights reserved
#


# externals
import weakref
# support
import pyre
# my protocol
from .Specification import Specification
# my superclass
from .Node import Node


# class declaration
class Product(Node, implements=Specification, internal=True):
    """
    The base class for data products
    """


    # public data
    @property
    def pyre_stale(self):
        """
        Retrieve my status
        """
        # delegate to my status manager
        return self.pyre_status.stale

    @pyre_stale.setter
    def pyre_stale(self, value):
        """
        Set my status
        """
        # delegate to my status manager
        self.pyre_status.stale = value
        # all done
        return


    # protocol obligations
    @pyre.export
    def pyre_make(self):
        """
        Invoke my factories to update me
        """
        # if i am not stale
        if not self.pyre_stale:
            # nothing to do
            return self
        # otherwise, go through my factories
        for factory in self.pyre_factories:
            # and ask each one to update me
            factory.pyre_make(product=self)
        # if all went well, update my status
        self.pyre_stale = False
        # all done
        return self


    def sync(self):
        """
        Examine my state
        """


    # meta-methods
    def __init__(self, **kwds):
        # chain up
        super().__init__(**kwds)
        # initialize the list of my factories
        self.pyre_factories = weakref.WeakSet()
        # all done
        return


    # flow hooks
    def pyre_newStatus(self, **kwds):
        """
        Build a handler for my status changes
        """
        # grab the factory
        from .ProductStatus import ProductStatus
        # make one and return it
        return ProductStatus(**kwds)


    def pyre_addInputBinding(self, factory):
        """
        Bind me as an input to the given {factory}
        """
        # let my monitor know there is a new client {factory}
        self.pyre_status.addInputBinding(factory=factory, product=self)
        # all done
        return


    def pyre_removeInputBinding(self, factory):
        """
        Unbind me as an input to the given {factory}
        """
        # let my monitor know there is a new client {factory}
        self.pyre_status.removeInputBinding(factory=factory, product=self)
        # all done
        return


    def pyre_addOutputBinding(self, factory):
        """
        Bind me as an output to the given {factory}
        """
        # add {factory} to my pile
        self.pyre_factories.add(factory)
        # let my monitor know there is a new client {factory}
        self.pyre_status.addOutputBinding(factory=factory, product=self)
        # all done
        return


    def pyre_removeOutputBinding(self, factory):
        """
        Unbind me as an output to the given {factory}
        """
        # remove {factory} from my pile
        self.pyre_factories.remove(factory)
        # let my monitor know there is a new client {factory}
        self.pyre_status.removeOutputBinding(factory=factory, product=self)
        # all done
        return


    # debugging support
    def pyre_dump(self, channel, indent=''):
        """
        Put some useful info about me in {channel}
        """
        # sign on
        channel.line(f"{indent}{self}")

        # my factories
        channel.line(f"{indent*2}factories:")
        for factory in self.pyre_factories:
            channel.line(f"{indent*3}{factory}")

        # my state
        channel.line(f"{indent*2}stale: {self.pyre_stale}")

        # my status monitor
        channel.line(f"{indent*2}status: {self.pyre_status}")
        channel.line(f"{indent*3}observers:")
        for observer in self.pyre_status.observers:
            channel.line(f"{indent*4}{observer}")

        # all done
        return self


# end of file

File: lib/pyre/flow/protocol/Product.cc
// -*- c++ -*-
//
// michael a.g. aïvázis <michael.aivazis@para-sim.com>
// (c) 1998-2024 all rights reserved

// support
#include "../public.h"

// destructor
pyre::flow::protocol::Product::~Product()
{
    // make a channel
    auto channel = pyre::journal::debug_t("pyre.flow.products.destroy");
    // show me
    channel
        // the product
        << "product " << this << ": destroy"
        << pyre::journal::newline
        // flush
        << pyre::journal::endl(__HERE__);
    // all done
    return;
}

// bindings
auto
pyre::flow::protocol::Product::addReader(name_type slot, factory_ref_type factory)
    -> product_ref_type
{
    // add the factory to my pile of readers
    _readers.insert({ slot, factory });
    // return a reference to me
    return ref();
};

auto
pyre::flow::protocol::Product::addWriter(name_type slot, factory_ref_type factory)
    -> product_ref_type
{
    // add the factory to my pile of writers
    _writers.insert({ slot, factory });
    // invalidate me
    flush();
    // return a reference to me
    return ref();
};

auto
pyre::flow::protocol::Product::removeReader(name_type slot, factory_ref_type factory)
    -> product_ref_type
{
    // remove the factory from my pile of readers
    _readers.extract({ slot, factory });
    // return a reference to me
    return ref();
};

auto
pyre::flow::protocol::Product::removeWriter(name_type slot, factory_ref_type factory)
    -> product_ref_type
{
    // remove the factory from my pile of writers
    _writers.extract({ slot, factory });
    // return a reference to me
    return ref();
};

// internals
auto
pyre::flow::protocol::Product::flush() -> void
{
    // make a channel
    auto channel = pyre::journal::debug_t("pyre.flow.products.flush");
    // show me
    channel
        // the product
        << "product '" << name() << "' at " << this << ": flush"
        << pyre::journal::newline
        // flush
        << pyre::journal::endl(__HERE__);
    // chain up
    Node::flush();
    // mark me
    dirty();
    // go through my readers
    for (auto & [slot, reader] : _readers) {
        // and flush each one
        reader->flush();
    }
    // all done
    return;
}

auto
pyre::flow::protocol::Product::make() -> product_ref_type
{
    // make a channel
    auto channel = pyre::journal::debug_t("pyre.flow.products.make");
    // show me
    channel
        // sign on
        << "product '" << name() << "' at " << this << ": make"
        << pyre::journal::newline
        // flush
        << pyre::journal::endl(__HERE__);

    // make a reference
    auto self = ref();
    // if i'm stale
    if (_stale) {
        // go through my writers
        for (auto & [slot, writer] : _writers) {
            // ask each one to refresh me
            writer->make(slot, self);
        }
        // mark me as clean
        clean();
    }
    // all done
    return self;
}

// end of file